README.md
# ChatGP-Terminal
<p align=justify>
  The simplest and cutest hahah C-program for querying ChatGPT from Linux terminal. At the moment, only model 'gpt-3.5-turbo' is supported. 
</p>
<ul>
  <li> Download: <a href="https://github.com/Lucho-A/ChatGP-Terminal/tree/master/ChatGP-Terminal/Releases">here</a></li>
  <li>Documentation (man): <a href="https://github.com/Lucho-A/ChatGP-Terminal/blob/master/ChatGP-Terminal/Releases/chatgp-terminal.1.gz">here</a></li>
</ul>
<p>
  Have Fun!
</p>

### Dependencies:
<ul>
  <li>libssl.so.3</li>
  <li>libreadline.so.8</li>
  <li>libc.so.6</li>
</ul>

You can install them by executing:


```
sudo apt-get install libssl3 libreadline8 libc6
```

### Usage:

```
./chatgp-terminal --help
```

### Screenshots:

![imagen](https://github.com/Lucho-A/ChatGP-Terminal/assets/40904281/5139e590-43c9-41b0-8426-c89bb7f7a397)

![imagen](https://github.com/Lucho-A/ChatGP-Terminal/assets/40904281/a0a8b580-c62a-493f-87e6-0bf5b6e7a03a)

![imagen](https://github.com/Lucho-A/ChatGP-Terminal/assets/40904281/54f0a77f-e340-4fc4-a020-9793b5aebd0f)




ChatGP-Terminal/CHANGELOG.md
#### v1.0.3:20230719

Bugs-Fixed:

- (I think) prompt (user's input) & readline() prompt solved
- Others like '%' on output

Others:

- default receiving message timeouting increased (35s)
- minor changes & code cleaned-up

#### v1.0.2:20230719

Improvements:

- Accepts as option '--show-finished-status'. If set, shows the status at the end of the answers.
	
Bugs-Fixed:

- Some I don't remember hhaha

Others:

- Show 'errno' description if getting info errors occurs.
- minor changes & code cleaned-up

#### v1.0.1:20230719

Improvements:

- allowing cancel by user handling signal ('SIGINT').
	
Bugs-Fixed:

- a lot (string parsings, data validations, error messages, not exit if errors, etc.).

Bugs-Known:

- readline() has some troubles for handling "e[0;XXm". I replaced #defines by "\033[0;XXm" in order to evaluate if behavior changes.
- JSON content message errors because absence of validation and parsing of user's messages ('?','\', etc). 

#### v1.0.0:20230718

- First version.


ChatGP-Terminal/Src/libGPT/libGPT.c
/*
 ============================================================================
 Name        : libGPT.c
 Author      : L. (lucho-a.github.io)
 Version     : 1.0.2
 Created on	 : 2023/07/18
 Copyright   : GNU General Public License v3.0
 Description : C file
 ============================================================================
*/

#include "libGPT.h"

#include <stdio.h>
#include <sys/socket.h>
#include <poll.h>
#include <openssl/ssl.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>

#define OPENAI_API_URL						"api.openai.com"
#define OPENAI_API_PORT						443

#define SOCKET_CONNECT_TIMEOUT_S			5
#define SOCKET_SEND_TIMEOUT_MS				5000
#define SOCKET_RECV_TIMEOUT_MS				60000

#define MAX_ROLES_SIZE						(1024*128)
#define BUFFER_SIZE_16K						(1024*16)

typedef struct Messages{
	char *userMessage;
	char *assistantMessage;
	struct Messages *nextMessage;
}Messages;

static struct Messages *historyContext=NULL;
static int contHistoryContext=0;
static int maxHistoryContext=0;

void libGPT_flush_history(void){
	while(historyContext!=NULL){
		Messages *temp=historyContext;
		historyContext=temp->nextMessage;
		if(temp->userMessage!=NULL) free(temp->userMessage);
		if(temp->assistantMessage!=NULL) free(temp->assistantMessage);
		free(temp);
	}
	historyContext=NULL;
	contHistoryContext=0;
}

int libGPT_clean(ChatGPT *cgpt){
	if(cgpt->api!=NULL) free(cgpt->api);
	if(cgpt->systemRole!=NULL) free(cgpt->systemRole);
	libGPT_flush_history();
	return RETURN_OK;
}

int libGPT_clean_response(ChatGPTResponse *cgptResponse){
	if(cgptResponse->httpResponse!=NULL) free(cgptResponse->httpResponse);
	if(cgptResponse->message!=NULL) free(cgptResponse->message);
	if(cgptResponse->finishReason!=NULL) free(cgptResponse->finishReason);
	return RETURN_OK;
}

int libGPT_init(ChatGPT *cgtp, char *api, char *systemRole, long int maxTokens, double temperature, int maxContextMessage){
	SSL_library_init();
	if(systemRole==NULL) systemRole=LIBGPT_DEFAULT_ROLE;
	if(maxTokens==0) maxTokens=LIBGPT_DEFAULT_MAX_TOKENS;
	if(temperature==0) temperature=LIBGPT_DEFAULT_TEMPERATURE;
	if(maxContextMessage==0) maxContextMessage=LIBGPT_DEFAULT_MAX_CONTEXT_MSGS;
	cgtp->api=malloc(strlen(api)+1);
	if(cgtp->api ==NULL) return LIBGPT_INIT_ERROR;
	snprintf(cgtp->api,strlen(api)+1,"%s",api);
	cgtp->systemRole=malloc(strlen(systemRole)+1);
	if(cgtp->systemRole ==NULL) return LIBGPT_INIT_ERROR;
	snprintf(cgtp->systemRole,strlen(systemRole)+1,"%s",systemRole);
	cgtp->maxTokens=maxTokens;
	cgtp->temperature=temperature;
	maxHistoryContext=maxContextMessage;
	return RETURN_OK;
}

int get_string_from_token(char *text, char *token, char *result, char endChar){
	int cont=0;
	char *message=strstr(text,token);
	for(int i=strlen(token);(message[i-1]=='\\' || message[i]!=endChar) && cont<BUFFER_SIZE_16K;i++,cont++) result[cont]=message[i];
	result[cont]='\0';
	if(cont>=BUFFER_SIZE_16K) return LIBGPT_BUFFERSIZE_OVERFLOW;
	return RETURN_OK;
}

static int libGPT_parse_result(char *messageSent, ChatGPTResponse *cgptResponse, bool createContext){
	char buffer[BUFFER_SIZE_16K]="";
	if(strstr(cgptResponse->httpResponse,"\"error\": {")!=NULL){
		if(get_string_from_token(cgptResponse->httpResponse,"\"message\": \"" ,buffer,'\"')==LIBGPT_BUFFERSIZE_OVERFLOW) return LIBGPT_BUFFERSIZE_OVERFLOW;
		cgptResponse->message=malloc(strlen(buffer)+1);
		snprintf(cgptResponse->message,strlen(buffer)+1,"%s", buffer);
		return LIBGPT_RESPONSE_MESSAGE_ERROR;
	}

	if(get_string_from_token(cgptResponse->httpResponse,"\"content\": \"",buffer,'\"')==LIBGPT_BUFFERSIZE_OVERFLOW) return LIBGPT_BUFFERSIZE_OVERFLOW;
	cgptResponse->message=malloc(strlen(buffer)+1);
	snprintf(cgptResponse->message,strlen(buffer)+1,"%s", buffer);

	if(createContext){
		Messages *newMessage=malloc(sizeof(Messages));
		newMessage->userMessage=malloc(strlen(messageSent)+1);
		snprintf(newMessage->userMessage,strlen(messageSent)+1,"%s",messageSent);
		free(messageSent);
		newMessage->assistantMessage=malloc(strlen(cgptResponse->message)+1);
		snprintf(newMessage->assistantMessage,strlen(cgptResponse->message)+1,"%s",cgptResponse->message);
		if(historyContext!=NULL){
			if(contHistoryContext>=maxHistoryContext){
				Messages *temp=historyContext->nextMessage;
				if(historyContext->userMessage!=NULL) free(historyContext->userMessage);
				if(historyContext->assistantMessage!=NULL) free(historyContext->assistantMessage);
				free(historyContext);
				historyContext=temp;
			}
			Messages *temp=historyContext;
			while(temp->nextMessage!=NULL) temp=temp->nextMessage;
			temp->nextMessage=newMessage;
		}else{
			historyContext=newMessage;
		}
		newMessage->nextMessage=NULL;
		contHistoryContext++;
	}

	if(get_string_from_token(cgptResponse->httpResponse,"\"finish_reason\": \"",buffer,'\"')==LIBGPT_BUFFERSIZE_OVERFLOW) return LIBGPT_BUFFERSIZE_OVERFLOW;
	cgptResponse->finishReason=malloc(strlen(buffer)+1);
	snprintf(cgptResponse->finishReason,strlen(buffer)+1,"%s", buffer);

	if(get_string_from_token(cgptResponse->httpResponse,"\"prompt_tokens\": ",buffer,'\n')==LIBGPT_BUFFERSIZE_OVERFLOW) return LIBGPT_BUFFERSIZE_OVERFLOW;
	cgptResponse->promptTokens=strtol(buffer,NULL,10);

	if(get_string_from_token(cgptResponse->httpResponse,"\"completion_tokens\": ",buffer,'\n')==LIBGPT_BUFFERSIZE_OVERFLOW) return LIBGPT_BUFFERSIZE_OVERFLOW;
	cgptResponse->completionTokens=strtol(buffer,NULL,10);

	if(get_string_from_token(cgptResponse->httpResponse,"\"total_tokens\": ",buffer,'\n')==LIBGPT_BUFFERSIZE_OVERFLOW) return LIBGPT_BUFFERSIZE_OVERFLOW;
	cgptResponse->totalTokens=strtol(buffer,NULL,10);

	return RETURN_OK;
}

int libGPT_send_chat(ChatGPT cgpt, ChatGPTResponse *cgptResponse, char *message, bool createContext){
	cgptResponse->httpResponse=NULL;
	cgptResponse->message=NULL;
	cgptResponse->finishReason=NULL;
	cgptResponse->promptTokens=0;
	cgptResponse->completionTokens=0;
	cgptResponse->totalTokens=0;
	char *messageParsed=malloc(strlen(message)*2);
	memset(messageParsed,0,strlen(message)*2);
	int cont=0;
	for(int i=0;i<strlen(message);i++,cont++){
		switch(message[i]){
		case '\"':
		case '\\':
			messageParsed[cont]='\\';
			messageParsed[++cont]=message[i];
			break;
		case '\n':
			messageParsed[cont]='\\';
			messageParsed[++cont]='n';
			break;
		default:
			messageParsed[cont]=message[i];
		}
	}
	messageParsed[cont]='\0';
	char roles[MAX_ROLES_SIZE]="", buf[BUFFER_SIZE_16K]="";
	Messages *temp=historyContext;
	while(temp!=NULL){
		snprintf(buf,BUFFER_SIZE_16K,"{\"role\":\"user\",\"content\":\"%s\"},",temp->userMessage);
		strcat(roles,buf);
		snprintf(buf,BUFFER_SIZE_16K,"{\"role\":\"assistant\",\"content\":\"%s\"},",temp->assistantMessage);
		strcat(roles,buf);
		temp=temp->nextMessage;
	}
	snprintf(buf,BUFFER_SIZE_16K,"{\"role\":\"user\",\"content\":\"%s\"}",messageParsed);
	strcat(roles,buf);
	char *payload=malloc(strlen(roles)+512);
	memset(payload,0,strlen(roles)+512);
	snprintf(payload,strlen(roles)+512,
			"{"
			"\"model\":\"gpt-3.5-turbo\","
			"\"messages\":"
			"["
			"{\"role\":\"system\",\"content\":\"%s\"},"
			"%s"
			"],"
			"\"max_tokens\": %ld,"
			"\"temperature\": %.2f"
			"}",cgpt.systemRole,roles,cgpt.maxTokens,cgpt.temperature);
	char *httpMsg=malloc(strlen(payload)+512);
	memset(httpMsg,0,strlen(payload)+512);
	snprintf(httpMsg,strlen(payload)+512,
			"POST /v1/chat/completions HTTP/1.1\r\n"
			"Host: %s\r\n"
			"user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64)\r\n"
			"accept: */*\r\n"
			"content-type: application/json\r\n"
			"authorization: Bearer %s\r\n"
			"connection: close\r\n"
			"content-length: %ld\r\n\r\n"
			"%s",OPENAI_API_URL,cgpt.api,strlen(payload),payload);
	free(payload);
	struct pollfd pfds[1];
	int numEvents=0,pollinHappened=0,bytesSent=0;
	SSL *sslConn=NULL;
	SSL_CTX *sslCtx=NULL;
	int socketConn=0;
	struct addrinfo hints, *res;
	memset(&hints, 0, sizeof(hints));
	hints.ai_family = AF_INET;
	hints.ai_socktype = SOCK_STREAM;
	if (getaddrinfo(OPENAI_API_URL, NULL, &hints, &res)!=0) return LIBGPT_GETTING_HOST_INFO_ERROR;
	struct sockaddr_in *ipv4 = (struct sockaddr_in *)res->ai_addr;
	void *addr = &(ipv4->sin_addr);
	char chatGptIp[INET_ADDRSTRLEN];
	inet_ntop(res->ai_family, addr, chatGptIp, sizeof(chatGptIp));
	freeaddrinfo(res);
	struct sockaddr_in serverAddress;
	serverAddress.sin_family=AF_INET;
	serverAddress.sin_port=htons(OPENAI_API_PORT);
	serverAddress.sin_addr.s_addr=inet_addr(chatGptIp);
	if((socketConn=socket(AF_INET, SOCK_STREAM, 0))<0) return LIBGPT_SOCKET_CREATION_ERROR;
	int socketFlags = fcntl(socketConn, F_GETFL, 0);
	fcntl(socketConn, F_SETFL, socketFlags | O_NONBLOCK);
	connect(socketConn, (struct sockaddr *) &serverAddress, sizeof(serverAddress));
	fd_set rFdset, wFdset;
	struct timeval tv;
	FD_ZERO(&rFdset);
	FD_SET(socketConn, &rFdset);
	wFdset=rFdset;
	tv.tv_sec=SOCKET_CONNECT_TIMEOUT_S;
	tv.tv_usec=0;
	if(select(socketConn+1,&rFdset,&wFdset,NULL,&tv)<=0) return LIBGPT_SOCKET_CONNECTION_TIMEOUT_ERROR;
	fcntl(socketConn, F_SETFL, socketFlags);
	if((sslCtx = SSL_CTX_new(SSLv23_method()))==NULL) return LIBGPT_SSL_CONTEXT_ERROR;
	if((sslConn = SSL_new(sslCtx))==NULL) return LIBGPT_SSL_CONTEXT_ERROR;
	if(!SSL_set_fd(sslConn, socketConn)) return LIBGPT_SSL_FD_ERROR;
	SSL_set_connect_state(sslConn);
	SSL_set_tlsext_host_name(sslConn, OPENAI_API_URL);
	if(!SSL_connect(sslConn)) return LIBGPT_SSL_CONNECT_ERROR;
	fcntl(socketConn, F_SETFL, O_NONBLOCK);
	pfds[0].fd=socketConn;
	pfds[0].events=POLLOUT;
	numEvents=poll(pfds,1,SOCKET_SEND_TIMEOUT_MS);
	if(numEvents==0){
		close(socketConn);
		return LIBGPT_SOCKET_SEND_TIMEOUT_ERROR;
	}
	pollinHappened=pfds[0].revents & POLLOUT;
	if(pollinHappened){
		int totalBytesSent=0;
		while(totalBytesSent<strlen(httpMsg)){
			bytesSent=SSL_write(sslConn, httpMsg + totalBytesSent, strlen(httpMsg) - totalBytesSent);
			totalBytesSent+=bytesSent;
		}
		free(httpMsg);
		if(bytesSent<=0){
			close(socketConn);
			return LIBGPT_SENDING_PACKETS_ERROR;
		}
	}else{
		free(httpMsg);
		close(socketConn);
		return LIBGPT_POLLIN_ERROR;
	}
	int bytesReceived=0,contI=0, totalBytesReceived=0;
	pfds[0].events=POLLIN;
	char buffer[BUFFER_SIZE_16K]="", bufferHTTP[BUFFER_SIZE_16K]="";
	do{
		numEvents=poll(pfds, 1, SOCKET_RECV_TIMEOUT_MS);
		if(numEvents==0){
			close(socketConn);
			return LIBGPT_SOCKET_RECV_TIMEOUT_ERROR;
		}
		pollinHappened = pfds[0].revents & POLLIN;
		if (pollinHappened){
			bytesReceived=SSL_read(sslConn,buffer, BUFFER_SIZE_16K);
			if(bytesReceived>0){
				totalBytesReceived+=bytesReceived;
				for(int i=0; i<bytesReceived && cont < BUFFER_SIZE_16K; i++, contI++) bufferHTTP[contI]=buffer[i];
				if(cont>=BUFFER_SIZE_16K) return LIBGPT_BUFFERSIZE_OVERFLOW;
				continue;
			}
			if(bytesReceived==0){
				close(socketConn);
				break;
			}
			if(bytesReceived<0 && (errno==EAGAIN || errno==EWOULDBLOCK)) continue;
			if(bytesReceived<0 && (errno!=EAGAIN)){
				close(socketConn);
				return LIBGPT_RECEIVING_PACKETS_ERROR;
			}
		}else{
			return LIBGPT_POLLIN_ERROR;
		}
	}while(TRUE);
	if(totalBytesReceived==0) return LIBGPT_ZEROBYTESRECV_ERROR;
	cgptResponse->httpResponse=malloc(strlen(bufferHTTP)+1);
	snprintf(cgptResponse->httpResponse,strlen(bufferHTTP)+1,"%s",bufferHTTP);
	return libGPT_parse_result(messageParsed, cgptResponse, createContext);
}


ChatGP-Terminal/Src/libGPT/libGPT.h
/*
 ============================================================================
 Name        : libGPT.h
 Author      : L. (lucho-a.github.io)
 Version     : 1.0.2
 Created on	 : 2023/07/18
 Copyright   : GNU General Public License v3.0
 Description : Header file
 ============================================================================
*/

#ifndef HEADERS_LIBGPT_H_
#define HEADERS_LIBGPT_H_

#define RETURN_ERROR 						-1
#define RETURN_OK 							0

#define LIBGPT_NAME 						"libGPT"
#define LIBGPT_MAJOR_VERSION				"1"
#define LIBGPT_MINOR_VERSION				"0"
#define LIBGPT_MICRO_VERSION				"2"
#define LIBGPT_VERSION						PROGRAM_MAJOR_VERSION"."PROGRAM_MINOR_VERSION"."PROGRAM_MICRO_VERSION
#define LIBGPT_DESCRIPTION					"C Library for ChatGPT"

#define	LIBGPT_DEFAULT_ROLE					""
#define	LIBGPT_DEFAULT_MAX_TOKENS			512
#define	LIBGPT_DEFAULT_TEMPERATURE			0.5
#define	LIBGPT_DEFAULT_MAX_CONTEXT_MSGS		3

#define PRINT_DBG							printf("\n\e[0;31mWTFFFF\e[0m\n");

typedef enum{
	FALSE=0,
	TRUE
}bool;

enum errors{
	LIBGPT_INIT_ERROR=-50,
	LIBGPT_GETTING_HOST_INFO_ERROR,
	LIBGPT_SOCKET_CREATION_ERROR,
	LIBGPT_SOCKET_CONNECTION_TIMEOUT_ERROR,
	LIBGPT_SSL_CONTEXT_ERROR,
	LIBGPT_SSL_FD_ERROR,
	LIBGPT_SSL_CONNECT_ERROR,
	LIBGPT_SOCKET_SEND_TIMEOUT_ERROR,
	LIBGPT_SENDING_PACKETS_ERROR,
	LIBGPT_POLLIN_ERROR,
	LIBGPT_SOCKET_RECV_TIMEOUT_ERROR,
	LIBGPT_RECEIVING_PACKETS_ERROR,
	LIBGPT_RESPONSE_MESSAGE_ERROR,
	LIBGPT_BUFFERSIZE_OVERFLOW,
	LIBGPT_ZEROBYTESRECV_ERROR
};

typedef struct ChatGPT{
	char *api;
	char *systemRole;
	long int maxTokens;
	double temperature;
}ChatGPT;

typedef struct ChatGPTResponse{
	char *httpResponse;
	char *message;
	char *finishReason;
	unsigned int promptTokens;
	unsigned int completionTokens;
	unsigned int totalTokens;
}ChatGPTResponse;

int libGPT_init(ChatGPT *, char *, char *, long int, double, int);
int libGPT_send_chat(ChatGPT,ChatGPTResponse *, char *, bool);
void libGPT_flush_history(void);
int libGPT_clean(ChatGPT *);
int libGPT_clean_response(ChatGPTResponse *);

#endif /* HEADERS_LIBGPT_H_ */


ChatGP-Terminal/Src/ChatGP-Terminal.c
/*
 ============================================================================
 Name        : ChatGP-Terminal.c
 Author      : L. (lucho-a.github.io)
 Version     : 1.0.9
 Created on	 : 2023/07/18
 Copyright   : GNU General Public License v3.0
 Description : Main file
 ============================================================================
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <readline/readline.h>
#include <readline/history.h>
#include <signal.h>
#include <errno.h>

#include "libGPT/libGPT.h"

#define PROGRAM_NAME				"ChatGP-Terminal"
#define PROGRAM_VERSION				"1.0.9"
#define PROGRAM_URL					"https://github.com/lucho-a/chatgp-terminal"
#define PROGRAM_CONTACT				"<https://lucho-a.github.io/>"

#define C_HRED 						"\e[0;91m"
#define C_RED 						"\e[0;31m"
#define C_HGREEN 					"\e[0;92m"
#define C_YELLOW 					"\e[0;33m"
#define C_HCYAN 					"\e[0;96m"
#define C_CYAN 						"\e[0;36m"
#define C_HWHITE 					"\e[0;97m"
#define C_WHITE 					"\e[0;37m"
#define C_DEFAULT 					"\e[0m"

#define PROMPT						";=exit) -> "
#define BANNER 						printf("\n%s%s v%s by L. <%s>%s\n\n",C_HWHITE,PROGRAM_NAME, PROGRAM_VERSION,PROGRAM_URL,C_DEFAULT);

#define	DEFAULT_RESPONSE_VELOCITY	20000

bool cancel=FALSE;

char * get_readline(char *prompt, bool addHistory){
	char *lineRead=(char *)NULL;
	if(lineRead){
		free(lineRead);
		lineRead=(char *)NULL;
	}
	lineRead = readline(prompt);
	if(lineRead && *lineRead && addHistory) add_history(lineRead);
	return(lineRead);
}

void print_result(ChatGPTResponse *cgptResponse, long int responseVelocity, bool showFinishReason, bool showPromptTokens, bool showCompletionTokens, bool showTotalTokens){
	printf("%s\n",C_HWHITE);
	for(int i=0;cgptResponse->message[i]!=0 && !cancel;i++){
		usleep(rand()%responseVelocity + 20000);
		if(cgptResponse->message[i]=='\\'){
			switch(cgptResponse->message[i+1]){
			case 'n':
				printf("\n");
				break;
			case 'r':
				printf("\r");
				break;
			case 't':
				printf("\t");
				break;
			case '\\':
				printf("\\");
				break;
			case '"':
				printf("\"");
				break;
			default:
				break;
			}
			i++;
			continue;
		}
		printf("%c",cgptResponse->message[i]);
		fflush(stdout);
	}
	printf("%s\n",C_DEFAULT);
	if(showFinishReason && !cancel) printf("\n%sFinish status: %s%s\n",C_DEFAULT,C_YELLOW,cgptResponse->finishReason);
	if(showFinishReason && cancel) printf("%s\nFinish status: %scanceled by user\n",C_DEFAULT,C_YELLOW);
	if(showPromptTokens) printf("%s\nPrompt tokens: %s%d\n",C_DEFAULT,C_YELLOW, cgptResponse->promptTokens);
	if(showCompletionTokens) printf("%s\nCompletion tokens: %s%d\n",C_DEFAULT,C_YELLOW, cgptResponse->completionTokens);
	if(showTotalTokens) printf("%s\nTotal tokens: %s%d\n",C_DEFAULT,C_YELLOW, cgptResponse->totalTokens);
	printf("%s\n",C_DEFAULT);
}

void usage(char *programName){
	BANNER;
	printf("See: man chatgp-terminal or refer to the documentation <https://rb.gy/75nbf>\n\n");
}

void signal_handler(int signalType){
	printf("\b\b  %s",C_DEFAULT);
	switch(signalType){
	case SIGINT:
		cancel=TRUE;
		break;
	default:
		break;
	}
}

void send_chat(ChatGPT cgpt, char *message, long int responseVelocity, bool showFinishedStatus, bool showPromptTokens, bool showCompletionTokens, bool showTotalTokens, bool createContext){
	ChatGPTResponse cgptResponse;
	int resp=0;
	if((resp=libGPT_send_chat(cgpt, &cgptResponse, message,createContext))!=RETURN_OK){
		switch(resp){
		case LIBGPT_ZEROBYTESRECV_ERROR:
			printf("\n%sOps, zero bytes received. Try again...%s\n\n",C_HRED,C_DEFAULT);
			break;
		case LIBGPT_BUFFERSIZE_OVERFLOW:
			printf("\n%sBuffer size exceeded. Pls, let me know <%s>.%s\n\n",C_HRED,PROGRAM_URL,C_DEFAULT);
			break;
		case LIBGPT_RESPONSE_MESSAGE_ERROR:
			printf("\n%s%s%s\n\n",C_HRED,cgptResponse.message,C_DEFAULT);
			break;
		case LIBGPT_SOCKET_RECV_TIMEOUT_ERROR:
			printf("\n%sTime out. Try again...%s\n\n",C_HRED,C_DEFAULT);
			break;
		default:
			if(cancel){
				printf("\n");
				break;
			}
			printf("\n%sDEBUG: Return error: %d. Errno: %d (%s).%s\n\n",C_HRED,resp, errno,strerror(errno),C_DEFAULT);
			break;
		}
	}else{
		print_result(&cgptResponse,responseVelocity, showFinishedStatus, showPromptTokens, showCompletionTokens, showTotalTokens);
	}
	libGPT_clean_response(&cgptResponse);
}

int main(int argc, char *argv[]) {
	signal(SIGINT, signal_handler);
	char *apikey="", *role=LIBGPT_DEFAULT_ROLE, *message=NULL;
	size_t len=0;
	int maxTokens=LIBGPT_DEFAULT_MAX_TOKENS, responseVelocity=DEFAULT_RESPONSE_VELOCITY, maxContextMessages=LIBGPT_DEFAULT_MAX_CONTEXT_MSGS;
	bool showFinishedStatus=FALSE,showPromptTokens=FALSE,showCompletionTokens=FALSE,showTotalTokens=FALSE, createContext=TRUE;
	double temperature=LIBGPT_DEFAULT_TEMPERATURE;
	for(int i=1;i<argc;i+=2){
		char *tail=NULL;
		if(strcmp(argv[i],"--version")==0){
			BANNER;
			exit(EXIT_SUCCESS);
		}
		if(strcmp(argv[i],"--apikeyfile")==0){
			FILE *f=fopen(argv[i+1],"r");
			if(f==NULL){
				printf("\n%sError opening API key file. Error %d: %s%s\n\n",C_HRED,errno,strerror(errno),C_DEFAULT);
				exit(EXIT_FAILURE);
			}
			while(getline(&apikey, &len, f)!=-1);
			apikey[strlen(apikey)-1]=0;
			fclose(f);
			continue;
		}
		if(strcmp(argv[i],"--apikey")==0){
			apikey=argv[i+1];
			continue;
		}
		if(strcmp(argv[i],"--role")==0){
			role=argv[i+1];
			continue;
		}
		if(strcmp(argv[i],"--max-tokens")==0){
			maxTokens=strtoul(argv[i+1],&tail,10);
			if(maxTokens<=0 || *tail!='\0'){
				printf("\n%sMax.Tokens value not valid.%s\n\n",C_HRED,C_DEFAULT);
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if(strcmp(argv[i],"--temperature")==0){
			temperature=strtod(argv[i+1],&tail);
			if(temperature<=0.0 || *tail!='\0'){
				printf("\n%sTemperature value not valid.%s\n\n",C_HRED,C_DEFAULT);
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if(strcmp(argv[i],"--response-velocity")==0){
			responseVelocity=strtod(argv[i+1],&tail);
			if(responseVelocity<=0 || *tail!='\0'){
				printf("\n%sResponse velocity value not valid.%s\n\n",C_HRED,C_DEFAULT);
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if(strcmp(argv[i],"--max-context-messages")==0){
			maxContextMessages=strtod(argv[i+1],&tail);
			if(maxContextMessages<=0 || *tail!='\0'){
				printf("\n%sMax. Context Messages value not valid.%s\n\n",C_HRED,C_DEFAULT);
				exit(EXIT_FAILURE);
			}
			continue;
		}
		if(strcmp(argv[i],"--message")==0){
			message=argv[i+1];
			continue;
		}
		if(strcmp(argv[i],"--show-finished-status")==0){
			showFinishedStatus=TRUE;
			i--;
			continue;
		}
		if(strcmp(argv[i],"--show-prompt-tokens")==0){
			showPromptTokens=TRUE;
			i--;
			continue;
		}
		if(strcmp(argv[i],"--show-completion-tokens")==0){
			showCompletionTokens=TRUE;
			i--;
			continue;
		}
		if(strcmp(argv[i],"--show-total-tokens")==0){
			showTotalTokens=TRUE;
			i--;
			continue;
		}
		if(strcmp(argv[i],"--no-create-context")==0){
			createContext=FALSE;
			i--;
			continue;
		}
		if(strcmp(argv[i],"--help")==0){
			usage(argv[0]);
			exit(EXIT_FAILURE);
		}
		usage(argv[0]);
		printf("%sArgument '%s' not recognized.%s\n\n",C_HRED,argv[i],C_DEFAULT);
		exit(EXIT_FAILURE);
	}
	ChatGPT cgpt;
	if(libGPT_init(&cgpt, apikey, role, maxTokens, temperature, maxContextMessages)==LIBGPT_INIT_ERROR){
		printf("\n%sError init structure. %s%s\n\n",C_HRED,strerror(errno),C_DEFAULT);
		exit(EXIT_FAILURE);
	}
	if(message!=NULL){
		send_chat(cgpt, message, responseVelocity, showFinishedStatus, showPromptTokens, showCompletionTokens, showTotalTokens, FALSE);
		libGPT_clean(&cgpt);
		exit(EXIT_SUCCESS);
	}
	rl_initialize();
	printf("\n");
	do{
		cancel=FALSE;
		printf("%s",C_HCYAN);
		char *message=get_readline(PROMPT, TRUE);
		printf("%s",C_DEFAULT);
		if(strcmp(message,";")==0) break;
		if(strcmp(message,"flush;")==0){
			libGPT_flush_history();
			printf("\n");
			continue;
		}
		if(strcmp(message,"")==0){
			printf("\n");
			continue;
		}
		send_chat(cgpt, message, responseVelocity, showFinishedStatus, showPromptTokens, showCompletionTokens, showTotalTokens, createContext);
	}while(TRUE);
	libGPT_clean(&cgpt);
	printf("\n");
	exit(EXIT_SUCCESS);
}
